### DISCUSSION TITLE
Understanding mock.restoreAll() and mock.reset() in Node.js Test Runner

### DISCUSSION THREAD

**Discussion by iatsiuk** [nodejs/help#4639] [Q&A] [answered]
I wont to ensure that the call count is correctly reset between test cases when using the Node.js Test Runner.

Here is my code:
```js
const assert = require("node:assert/strict");
const { describe, afterEach, it, mock } = require("node:test");

describe("", async () => {
  let testFn = mock.fn();

  afterEach(() => {
    mock.reset();
    // or should we use mock.restoreAll() instead?
  });

  it("", () => {
    testFn();
  });

  it("", () => {
    testFn();

    assert.strictEqual(testFn.mock.calls.length, 1);
  });
});
```

Using `mock.reset()` should theoretically reset the call count for the `testFn` mock between test cases. However, it seems that the Node.js Test Runner's mock implementation may have a bug or behave differently than expected in this case:
```
    AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

    2 !== 1
```

Here is another example, that using sinon.js `fake()`:
```js
const assert = require("node:assert/strict");
const { describe, afterEach, it } = require("node:test");
const sinon = require("sinon");

describe("", async () => {
  let testFn = sinon.fake();

  afterEach(() => {
    sinon.reset();
  });

  it("", () => {
    testFn();
  });

  it("", () => {
    testFn();

    assert.strictEqual(testFn.callCount, 1);
  });
});
```
In the second code example, using `sinon.reset()` from the Sinon library correctly resets the call count for the `testFn` fake between test cases.

How to correctly reset the number of mock calls using the Node.js test runner. What should mock.restoreAll() and mock.reset() be used for?

**Comment by cjihrig** [comment, marked as answer]
You probably want `testFn.mock.resetCalls();` instead of `mock.reset();`

**Comment by iatsiuk** [comment]
@cjihrig Thank you for response!

I understand that it is possible to reset the state of each mock individually or even re-assign the variable that holds it. My question was related to the cases where it is appropriate to use mock.restoreAll() and mock.reset() from the global scope. I expect that calling these methods SHOULD reset the call count of the mock, otherwise, what other purpose do they serve?

**Reply by cjihrig** [reply]
`testFn.mock` is an instance of [`MockFunctionContext`](https://nodejs.org/api/test.html#class-mockfunctioncontext). Each individual mocked function has one of these, and this is the only way to get information about that particular mock.

The top level `mock`, and the per test context `t.mock`, are instances of [`MockTracker`](https://nodejs.org/api/test.html#class-mocktracker). This is used to create and restore mocks. This manages mock functions, methods, modules, etc. Restoring a mock does not change the behavior of the `MockFunctionContext` instances. Instead, it makes sure that mocks are restored to their original behavior. That is why if you create mocks with `t.mock`, the test runner is able to unmock things after the test ends.
